vscode/src/language-service/testExplorer.ts

```diff
@@ -0,0 +1,220 @@
+      }
+    }
+  };
+
+  /**
+   * Given a single test case, run it in the worker (which runs the interpreter) and report results back to the
+   * `TestController` as a side effect.
+   *
+   * This function manages its own event target for the results of the test run and uses the controller to render the output in the VS Code UI.
+   **/
+  async function runTestCase(
+    ctrl: vscode.TestController,
+    testCase: vscode.TestItem,
+    request: vscode.TestRunRequest,
+    worker: ICompilerWorker,
+  ): Promise<void> {
+    log.trace("Running Q# test: ", testCase.id);
+    if (testCase.children.size > 0) {
+      for (const childTestCase of testCase.children) {
+        await runTestCase(ctrl, childTestCase[1], request, worker);
+      }
+      return;
+    }
+    const run = ctrl.createTestRun(request);
+    const evtTarget = createDebugConsoleEventTarget((msg) => {
+      run.appendOutput(`${msg}\n`);
+    });
+    evtTarget.addEventListener("Result", (msg) => {
+      if (msg.detail.success) {
+        run.passed(testCase);
+      } else {
+        const failureLocation =
+          msg.detail?.value?.uri ||
+          (msg.detail?.value?.related &&
+            msg.detail.value.related[0].location?.source) ||
+          null;
+
+        const message: vscode.TestMessage = {
+          message: msg.detail.value.message,
+          location:
+            failureLocation === null
+              ? undefined
+              : {
+                  range: toVsCodeRange(msg.detail.value.range),
+                  uri: vscode.Uri.parse(failureLocation),
+                },
+        };
+        run.failed(testCase, message);
+      }
+      run.end();
+    });
+
+    const callableExpr = `${testCase.id}()`;
+    const uri = testCase.uri;
+    if (!uri) {
+      log.error(`No compilation URI for test ${testCase.id}`);
+      run.appendOutput(`No compilation URI for test ${testCase.id}\r\n`);
+      return;
+    }
+    const programResult = await getProgramForDocument(uri);
+
+    log.info(JSON.stringify(programResult, null, 2));
+    if (!programResult.success) {
+      throw new Error(programResult.errorMsg);
+    }
+
+    const program = programResult.programConfig;
+
+    try {
+      await worker.run(program, callableExpr, 1, evtTarget);
+    } catch (error) {
+      log.error(`Error running test ${testCase.id}:`, error);
+      run.appendOutput(`Error running test ${testCase.id}: ${error}\r\n`);
+    }
+    log.trace("ran test:", testCase.id);
+  }
+
+  testController.createRunProfile(
+    "Interpreter",
+    vscode.TestRunProfileKind.Run,
+    runHandler,
+    true,
+    undefined,
+    false,
+  );
+
+  const testVersions = new WeakMap<vscode.TestItem, number>();
+  async function onTestCallables(evt: {
+    detail: {
+      callables: ITestDescriptor[];
+    };
+  }) {
+    let currentVersion = 0;
+    for (const [, testItem] of testController.items) {
+      currentVersion = (testVersions.get(testItem) || 0) + 1;
+      break;
```

Comment:
So this immediately breaks after the first item? Why use a for loop at all rather than just get the first item (or 1 if there isn't one)?

Reply:
Oh right, it's a WeakMap and there is no ordering... don't mind me.

Reply:
I did this because there's no index signature in TS for testcontroller.items:

<img width="1032" alt="image" src="https://github.com/user-attachments/assets/3ff2f941-23ec-436e-8749-c3c3d95df960" />

and `.get` wants a string key:
<img width="838" alt="image" src="https://github.com/user-attachments/assets/c34a9fa3-707a-4a69-801d-07f62c3c2ffd" />

but it _is_ an iterator. So this was how I figured I could get the first element while appeasing the API. Is there a better way?